### PR TITLE
[API] Add structure for tensor_trainer's sub-plugin @open sesame 12/27 11:50

### PR DESCRIPTION
**Pull request by songgot**
[API] Add structure for tensor_trainer's sub-plugin

- Create nnstreamer_plugin_api_trainer.h
- Add GstTensorTrainerFramework to define tensor_trainer sub-plugin
- Add GstTensorTrainerProperties to send tensor_trainer properties

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

Signed-off-by: hyunil park <hyunil46.park@samsung.com>

Related:
https://github.com/nnstreamer/nnstreamer/pull/3949
https://github.com/nnstreamer/nnstreamer/pull/3986
https://github.com/nnstreamer/nntrainer/pull/2059

### CODE REVIEW

File: gst/nnstreamer/include/nnstreamer_plugin_api_filter.h [hunk 1]
```diff
@@ -615,6 +615,84 @@ extern void
...
+typedef struct _GstTensorTrainerProperties
+{
+  const char *model_config;      /* The configuration file path for creating model */
+  unsigned int input_tensors_size[NNS_TENSOR_SIZE_LIMIT];       /* Size of tensor in byte for each input type */
+  unsigned int num_input_tensors;       /* The number of input tensors, The number of input tensors is equal to the number of inputs plus the number of labels. */
+  unsigned int num_inputs;      /* The number of input lists, the input is where framework receive the features to train the model, num_inputs indicates how many inputs there are. */
+  unsigned int num_labels;      /* The number of label lists, the label is where framework receive the class to train the model, num_labels indicates how many labels there are. */
+  unsigned int num_train_samples;       /* The number of train sample used to train the model. */
+  unsigned int num_valid_samples;       /* The number of valid sample used to train the model. */
+} GstTensorTrainerProperties;
```
Comment (myungjoo) [inline gst/nnstreamer/include/nnstreamer_plugin_api_filter.h:633]: For the tensor metadata, use nnstreamer standard data structure: i.e., ```GstTensorsInfo```

File: gst/nnstreamer/include/nnstreamer_plugin_api_filter.h [hunk 2]
```diff
@@ -615,6 +615,84 @@ extern void
...
+/* extern functions for subplugin management, exist in tensor_trainer.c */
+/**
+ * @brief Trainer's sub-plugin should call this function to register itself.
+ * @param[in] sub_plugin tensor_trainer sub-plugin to be registered.
+ * @return TRUE if registered. FALSE is failed or duplicated.
+ *
+ * @note Do not change the subplugins callbacks after probing the filter.
+ */
+extern int
+nnstreamer_trainer_probe (GstTensorTrainerFramework * sub_plugin);
```
Comment (myungjoo) [inline gst/nnstreamer/include/nnstreamer_plugin_api_filter.h:694]: constify sub_plugin param?

File: gst/nnstreamer/include/nnstreamer_plugin_api_filter.h (hunk 1 above)
Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_filter.h:633]: Thank you.
1. We can calculate the size of each tensors using GstTensorsInfo and gst_tensor_info_get_size(&info->info[index]).
Is it against policy to calculate size and pass it to sub-plugin?

File: gst/nnstreamer/include/nnstreamer_plugin_api_filter.h [hunk 3]
```diff
@@ -615,6 +615,84 @@ extern void
...
+   * tensor_trainer call this to send tensor data to subplugin
+   * @param[in] prop read-only property values
+   * @param[in/out] private_data, a subplugin may save its internal private data here.
+   * @param[in] input The array of input tensors. Allocated and filled by tensor_trainer
+   * @param[out] output The array of output tensors. Allocated by tensor_trainer, set to NULL if not used.
+   */
+  int (*invoke_NN) (const GstTensorTrainerProperties * prop,
+      void **private_data, const GstTensorMemory * input,
+      GstTensorMemory * output);
+
```
[the remark this answers is not in the document]
Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_filter.h:682]: Thank you.
I will check this.

File: gst/nnstreamer/include/nnstreamer_plugin_api_filter.h (hunk 2 above)
Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_filter.h:694]: thank you.
I will fix it like tensor_filter.

/* extern functions for subplugin management, exist in tensor_filter.c */
/**
 * @brief Filter's sub-plugin should call this function to register itself.
 * @param[in] tfsp Tensor-Filter Sub-Plugin to be registered.
 * @return TRUE if registered. FALSE is failed or duplicated.
 *
 * @note Do not change the subplugins callbacks after probing the filter.
 */
extern int
nnstreamer_filter_probe (GstTensorFilterFramework * tfsp);

File: gst/nnstreamer/include/nnstreamer_plugin_api_filter.h [hunk 4]
```diff
@@ -615,6 +615,84 @@ extern void
 nnstreamer_filter_shared_model_replace (void *instance, const char *key,
     void *new_interpreter, void (*replace_callback) (void *, void *), void (*free_callback) (void*));
 
+/**
```
[the remark this answers is not in the document]
Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_filter.h]: thank you. I was working on a move.
thank you.

**Comment by songgot** [comment]
Related:
https://github.com/nnstreamer/nnstreamer/pull/3949
https://github.com/nnstreamer/nntrainer/pull/2059

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+/**
+ * @brief GstTensorTrainer's properties for neural network framework (internal data structure)
+ *
+ * Internal data of GstTensorTrainer required by tensor_trainer's custom subplugin.
+ */
+typedef struct _GstTensorTrainerProperties
+{
+  GstTensorsInfo input_meta;      /**< configured input tensor info */
+  const char *model_config;     /* The configuration file path for creating model */
+  const char *model_save_path;      /* The file path to save the model */
```
Comment (myungjoo) [inline gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:32]: These two look too specialized for nntrainer. Make them general for all training frameworks. Be ready for model-format updates in nntrainer in the future, too.

Questions you need to ask for yourself:
1. what if a framework requires a single file for training?
2. what if a framework requires three files for training?
3. what if a framework supports multiple different formats (and different number of files accordingly)?
...

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief GstTensorTrainer's properties for neural network framework (internal data structure)
+ *
+ * Internal data of GstTensorTrainer required by tensor_trainer's custom subplugin.
+ */
```
Comment (myungjoo) [inline gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:27]: The each attribute in the struct, please use the correct Doxygen format.

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+ * @brief tensor_trainer Subplugin definition
+ *
+ * Common callback parameters:
+ * prop Trainer properties. Read Only.
+ * private_data Subplugin's private data. Set this (*private_data = XXX) if you want to change trainer->private_data.
+ */
+typedef struct _GstTensorTrainerFramework
+{
+  /* The name of the neural network, searchable by 'framework' property */
+  char *name;
```
Comment (myungjoo) [inline gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:49]: Please follow what GstTensorFilterFramework v1 did, not v0.

- Separation of properties: GstTensorFilterFrameworkInfo
- Separation of event handling: GstTensorFilterFrameworkEventData

This allows higher flexibility: allowing different configuations in each instance.

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h (hunk 6 above)
Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:27]: Thank you. I will fix.

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h (hunk 5 above)
Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:32]: Thank you.
About model_config, I've been thinking a lot. How about 'config_string' ?
In case of tensorflow (https://www.tensorflow.org/api_docs/python/tf/keras/models/model_from_json), use 'config' in example. param name is json_strring.
```
model = tf.keras.Sequential([
    tf.keras.layers.Dense(5, input_shape=(3,)),
    tf.keras.layers.Softmax()])
config = model.to_json()
loaded_model = tf.keras.models.model_from_json(config)
```
nntrainer use 'config'
```
  /**
   * @brief     Create and load the Network with ini configuration file.
   * @param[in] config config file path
   * @retval #ML_ERROR_NONE Successful.
   * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
   */
  virtual int loadFromConfig(const std::string &config) = 0;
```  
About questions. It seems that these can be solved with GstTensorsInfo.

Currently for test.
MNIST  is also a single file for training. so after separating the file, I am going to test it by muxing it with tensor_mux and put it into tensor_trainer.

The simplest way is that separate input and label are ped into the input of the pipeline and then mux them with tensor_mux.
But....
Considering what you said.
For Various types of data(a single file for training, ....), I think it's better to process data separation outside of tensor_trainer rather than inside it.
It would be nice to if fixed GstTensorsInfo and GstTensorMemory came into tensor_trainer. 

There must be a better way, but I can't think of it yet.
We will gradually develop this.

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+/**
+ * @brief tensor_trainer Subplugin definition
+ *
+ * Common callback parameters:
+ * prop Trainer properties. Read Only.
+ * private_data Subplugin's private data. Set this (*private_data = XXX) if you want to change trainer->private_data.
+ */
+typedef struct _GstTensorTrainerFramework
+{
+  /* The name of the neural network, searchable by 'framework' property */
```
[the remark this answers is not in the document]
Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:48]: Okay, Thank  you.

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h (hunk 7 above)
Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:49]: Okay, Thank you.
I will put what you said while adding features.

Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:49]: Thank you!!

**Review by myungjoo — APPROVED** [review]

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h [hunk 9]
```diff
@@ -0,0 +1,144 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
+/**
+ * NNStreamer API for Tensor_Trainer Sub-Plugins
+ * Copyright (C) 2022 Hyunil Park <hyunil46.park@samsung.com>
+ */
+/**
+ * @file  nnstreamer_plugin_api_filter.h
+ * @date  1 Dec 2022
+ * @brief Mandatory APIs for NNStreamer Trainer sub-plugins (No External Dependencies)
```
Comment (wooksong) [inline gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:9]: What does ```(No External Dependencies)``` mean precisely?

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+ *
+ * Internal data of GstTensorTrainer required by tensor_trainer's custom subplugin.
+ */
+typedef struct _GstTensorTrainerProperties
+{
+  GstTensorsInfo input_meta;    /**< configured input tensor info */
+  const char *model_config;    /**< The configuration file path for creating model */
+  const char *model_save_path;    /**< The file path to save the model */
+  int64_t num_inputs;    /**< The number of input lists, the input is where framework receive the features to train the model, num_inputs indicates how many inputs there are. */
+  int64_t num_labels;    /**< The number of label lists, the label is where framework receive the class to train the model, num_labels indicates how many labels there are. */
```
Comment (wooksong) [inline gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:37]: Is it common for all?

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * @brief tensor_trainer Subplugin definition
+ *
+ * Common callback parameters:
+ * prop Trainer properties. Read Only.
+ * private_data Subplugin's private data. Set this (*private_data = XXX) if you want to change trainer->private_data.
+ */
+struct _GstTensorTrainerFramework
+{
+  uint64_t version;
```
Comment (wooksong) [inline gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:65]: The format for the version looks the same as that of the tensor filter's version. Then, how about making it common, if possible?

File: packaging/nnstreamer.spec [hunk 12]
```diff
@@ -1134,6 +1134,7 @@ cp -r result %{buildroot}%{_datadir}/nnstreamer/unittest/
 %files single-devel
 %{_includedir}/nnstreamer/tensor_typedef.h
 %{_includedir}/nnstreamer/nnstreamer_plugin_api_filter.h
+%{_includedir}/nnstreamer/nnstreamer_plugin_api_trainer.h
```
Comment (wooksong) [inline packaging/nnstreamer.spec:1137]: Is it necessary to add this header to the RPM at this moment? IMO, at least, packaging should be handled as a separate commit.

Comment (songgot) [reply packaging/nnstreamer.spec:1137]: Thank you.
I will separate

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h (hunk 11 above)
Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:65]: Thank you. Good idea.

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h (hunk 10 above)
Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:37]: Thank you.
I can't say for sure now. This structure is still under development.
However, since this PR is related to nntrainer and nnstreamer tensor_trainer, merge is needed. That's why [myungjoo](https://github.com/myungjoo) approved.
This structure is a temporay code. And it seems that a lot will change in the future.
I'll make a new PR when I need to fix it.

**Comment by songgot** [comment]
Thank you.
I can't say for sure now. This structure is still under development.
However, since this PR is related to nntrainer and nnstreamer tensor_trainer, merge is needed. That's why [myungjoo](https://github.com/myungjoo) approved.
This structure is a temporay code. And it seems that a lot will change in the future.
I'll make a new PR when I need to fix it.

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h (hunk 10 above)
Comment (wooksong) [reply gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:37]: So are you saying ```num_labels``` is required to merge this PR? If not, you are trying to push the unnecessary line to upstream. It is better to leave the NECESSARY lines to be required by nntrainer.

Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:37]: Sorry. I thought what you mentioned was not about ```'num_labels```, but about all member variables.
We also need ```'num_labels```.
we can think of some case.
There can be more than one label for multiple features.
```
Feature = N, label = 2
feature 1----|
feature 2----|                             |--> label1
feature 3----| ---------> [DNN]----|--> label2 
...
feature N---|
```
For TTS support, nntrainer has received multiple labels.
Thank you.

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h (hunk 9 above)
Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:9]: This term is also used for tensor_plugin_api_filter.h
It seems that this header file has no external dependencies.
Thank you.

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h (hunk 11 above)
Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:65]: Taking this out and using it in common doesn't seem like something I can do right now.
I will do it with another PR when the policy and direction are decided.

File: packaging/nnstreamer.spec (hunk 12 above)
Comment (songgot) [reply packaging/nnstreamer.spec:1137]: Done, thank you ([03e3085](https://github.com/nnstreamer/nnstreamer/pull/3977/commits/03e30851427b383ca6f5b7bb0df7ac090a5a1869))

**Review by jaeyun-jung — APPROVED** [review]

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h [hunk 13]
```diff
@@ -0,0 +1,144 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
+/**
+ * NNStreamer API for Tensor_Trainer Sub-Plugins
+ * Copyright (C) 2022 Hyunil Park <hyunil46.park@samsung.com>
+ */
+/**
+ * @file  nnstreamer_plugin_api_filter.h
```
Comment (gichan-jang) [inline gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:7]:
```suggestion
 * @file  nnstreamer_plugin_api_trainer.h
```

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+      const GstTensorTrainerProperties * prop, void *private_data);
+  /**< tensor_trainer call this to train the model
+   * @param[in] prop read-only property values
+   * @param[in] private_data, a subplugin may save its internal private data here.
+   * @return 0 if ok. < 0 if error.
+   */
+
+  int (*invoke) (const GstTensorTrainerFramework * self,
+      const GstTensorTrainerProperties * prop,
+      void *private_data, const GstTensorMemory * input);
```
Comment (gichan-jang) [inline gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h:96]: Don't need to get the result?

File: gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h (hunk 13 above)
Comment (songgot) [reply gst/nnstreamer/include/nnstreamer_plugin_api_trainer.h]: Thank you. done([d90c74c](https://github.com/nnstreamer/nnstreamer/pull/3977/commits/d90c74c602f83f7a0ed1629925b2d49539b1c401))

**Review by gichan-jang — APPROVED** [review]
LGTM

**Comment by myungjoo** [comment]
You may keep updating this internal API as this is still in experimental phase.

**Comment by songgot** [comment]
okay, thank you.
